### PR TITLE
Generalize aggressive LICM for all order-independent loops with GPU locale model

### PR DESCRIPTION
**Pull request by e-kayrakli**
This logically reverts some of the GPU transforms changes in https://github.com/chapel-lang/chapel/pull/22664. Note that the code has evolved and got significantly complicated since, so this is not a simple reversion.

While there, this PR also removes the deprecated `setBlockSize` and removes some testing for bare `__primitive("set block size"` since with the attribute-based approach, we expect that primitive to appear in a special block, and as such it made sense to disallow it being used by the user directly.

The main purpose of the PR is to reduce some of the complications that revolves around generating GPU loop clones early to subject them to aggressive LICM. gpuTransforms still needs a lot of love in my view, but this is the first step towards a bigger change in how we outline loops.

The biggest challenge was to handle atrribute/primitive blocks for GPU eligible loops. Previously, where we clone loops early, we could handle such blocks differently as we can pretty much just consume them directly for GPU clones. Now that there's no GPU clone of the loop we need to be a bit more careful. To handle that:

- before we generate kernels, we generate "launch block"s. These blocks become the GPU branch of the conditional checking the sublocale.
- creating these early, we can now start populating it before we actually start generating the kernel. Note that kernel generation requires blockSize to be populated so that reductions can be handled. (There's a bug there that this PR does not address. See https://github.com/chapel-lang/chapel/issues/25739)
- we create the kernel normally after having created that stub
- then, finally we go back and finalize that block and the dynamic check. That step is mostly simplified and just about creating the kernel configuration, launching it and cleaning up its memory

Test:
- [x] nvidia
- [x] nvidia with gpu specialization (spot-check)
- [x] amd
- [x] cpu

### CODE REVIEW

**Comment by e-kayrakli** [comment]
I bumped into https://github.com/chapel-lang/chapel/issues/25739 while working on this. There are some recklessness in the implementation as to how we handle blockSize, which is expected to be an immediate by the codegen.

File: compiler/optimizations/gpuTransforms.cpp [hunk 1]
```diff
@@ -1962,87 +1930,69 @@ static VarSymbol* generateNumThreads(BlockStmt* gpuLaunchBlock,
...
+  }
 
-    gpuBlock->insertAtTail(primitivesBlock->remove());
-    primitivesBlock->flattenAndRemove();
+  // we are done with the call, now remove it.
+  if (blockSizeCall()) {
+    if (blockSizeCall()->inTree()) {
+      blockSizeCall()->remove();
+    }
   }
```
Comment (DanilaFe) [inline compiler/optimizations/gpuTransforms.cpp:1950]:
```suggestion
  if (auto bsc = blockSizeCall()) {
    if (bsc->inTree()) {
      bsc->remove();
    }
  }
```

**Review by DanilaFe — APPROVED** [review]

File: compiler/optimizations/gpuTransforms.cpp (hunk 1 above)
Comment (e-kayrakli) [reply compiler/optimizations/gpuTransforms.cpp]: https://github.com/chapel-lang/chapel/pull/25738/commits/ac83facb32ef0bba53974e928b9bab6e3c9b2927

**Comment by e-kayrakli** [comment]
I am re-running AMD testing again before merging this.

**Pull request merged**